TechnicalDocuments/Proposals/IdentifiersForUIAutomation.md

```diff
@@ -0,0 +1,203 @@
+    subscript<Renderer: BoxRenderer>(
+        _ renderer: Renderer.Type,
+        _ section: Renderer.SectionID,
+        _ item: Renderer.ItemID
+    ) -> String {
+        return "\(section)/\(item)/"
+    }
+
+    subscript<Renderer: BoxRenderer, AutomationName: RawRepresentable>(
+        _ renderer: Renderer.Type,
+        _ section: Renderer.SectionID,
+        _ item: Renderer.ItemID,
+        _ automationName: AutomationName
+    ) -> String where AutomationName.RawValue == String {
+        return "\(section)/\(item)/" + automationName.rawValue
+    }
+}
+```
+
+which the automation code can rely on. For example:
+
+```swift
+import BabylonHealthManagementUI
+
+extension BaseScreen {
+    // NOTE: alias in the UI testing target to help reduce verbosity.
+    typealias Name = DesignLibrary.AutomationName
+}
+
+class FeedScreen: BaseScreen {
+    func tapSubmit() {
+        let button = app[FeedRenderer.self, .activity, .card, Name.PrettyCard.submit]
+        button.tap()
+    }
+}
+```
+
+### Solving the associated enum issue
+
+In our use cases, it is very common to have section IDs and item IDs associated with unique identifiers of some other domain models, and hence the use of enums with associated values.
+
+Unfortunately, results from printing these enums via `String.init` is not quite ideal for our use cases. For example, considering this snippet:
+
+```swift
+enum SpecialItem {
+    case bento
+    case sushi
+}
+
+enum ItemID {
+    case item(SpecialItem)
+    case generic(String)
+}
+```
+
+Printing `.item(.bento)` results in `item(SwiftPlayground.SpecialItem.bento)`, which includes the fully qualified type name of the enum as part of the associated payload. On the other hand, printing `.generic("wasabi")` results in `generic("wasabi")`, which preserves the double quote.
+
+The issue can be resolved by implementing `CustomStringConvertible` on the ID enums. But to improve the developer experience, the proposal sugggests also providing a utility printing enums in a consistent way that works for us. Specifically, these criteria should be met:
+
+* enum without associated values should be printed as `caseName`; and
+* enum with associated values should be printed as `caseName(value)`; and
+* integer and strings as associated values should be printed as is without any punctuation added; and
+* Repeated printing should always lead to the same outcome.
+
+The utility would be in the form of a `AutoPrettyEnumDescription` protocol which refines `CustomStringConvertible` with a default implementation of `description`. It would use `Mirror` internally.
+
+A caveat: There is a scenario, filed as [SR-10272](https://bugs.swift.org/browse/SR-10272), where `Mirror` does not reflect an enum correctly as expected. But it should be rare in practice.
+
+## Impact on existing codebase
+Additive changes.
+
+## Alternatives considered
+Status quo: Fragile accessibility identifier; ad hoc definition of accessibility identifier.
```

Comment:
I think we can do better in finding alternative solutions, like in Swift evolution proposal, rather than just suggesting to do things in one particular way or leave it as it is.

Reply:
One of such solutions can be instead of relying on automatically generated accessibility identifiers to require components to have accessibility identifier parameter (even if just per convention) and to have these identifiers defined manually in a name space in a renderer, so that tests code can access it easily. This way they wouldn't depend on unique identifiers as for automation it's most of the time enough to query elements by their index, otherwise they can be still queried by their textual content if specific element need to be accessed.

```swift
class FeedRenderer {
  func render(...) -> Box {
    return .empty
             |-+ Section(id: .activity)
             |---+ Node(
                     id: .card, 
                     component: prettyCard(
                       accessibilityIdentifier: AccessibilityIDs.activityCard(1)
                     )
             )
             |---+ Node(
                     id: .submit,
                     component: button(
                       accessibilityIdentifier: AccessibilityIDs.submitButton
                     )
             )
  }
}

extension FeedRenderer {
  enum AccessibilityIDs {
    static let activityCard: (Int) -> String = { "activity_card_\($0)" }
    static let submitButton = "submit"
  }
}
```